# bacterial_metagenomes_pangraphs
A workflow for complete analysis of bacterial metagenomes and pagenome graphs and direct viewing in panchae. It will also analyze metagenomes from both illumina and long reads. If you want in nextflow then leave a comment and i will code in nextflow, currently it runs on slurm and pbs plus docker instances. if you are running this on the slurm or the pbs then also you can run it directly. Adding the support for the singularity containers. soon will push a docker image. 

updated 2023-10-19 with complete workflow and added support for snpcalling and others. next update will be the support for the multi mappers, graph creation, pangenome construction, evolutionary analysis, genome comparison, python  and R code will be intergated, plus ruby bindings and also a ruby on rails application for submission to the server plus docker release. 

This workflow is still in development and will be updated soon.

```
docker pull ubuntu-bacterial-pangraphs
```

Gaurav Sablok \
ORCID: https://orcid.org/0000-0002-4157-9405 \
WOS: https://www.webofscience.com/wos/author/record/C-5940-2014 \
RubyGems Published: https://rubygems.org/profiles/sablokgaurav \
Python Packages Published : https://pypi.org/user/sablokgaurav/
